### ISSUE TITLE
Some Chars are swallowed

### ISSUE THREAD

**Issue by MaxDaten** [Yuras/pdf-toolbox#24]
Hi,

thanks for the library!

But unfortunately I've bumped into a strange issue with the text extraction, maybe I missed something and you can help me:

This example:

```haskell
import System.IO
import Data.Text.IO as T
import Pdf.Toolbox.Document

main = withBinaryFile "input.pdf" ReadMode $ \handle ->
  runPdfWithHandle handle knownFilters $ do
    doc <- document
    catalog <- documentCatalog doc
    rootNode <- catalogPageNode catalog
    count <- pageNodeNKids rootNode
    liftIO $ print count
    page <- pageNodePageByNum rootNode 0
    liftIO . T.putStrLn =<< pageExtractText page
```

on this (german) PDF extracted page: https://drive.google.com/file/d/0B17tmmsfO8Pcc0VnLV9zRjlEODA/view?usp=sharing

gives this two lines (among other lines)

```plain
Erschrocken blicke ich an mir herab. atsächlich: Meine 
rme und Beine haben dieselbe arbe. 
``` 

But should be:

```
Erschrocken blicke ich an mir herab. Tatsächlich: Meine 
Arme und Beine haben dieselbe Farbe.
```

Any ideas whats happening? I tested it with other pdfs with similar issues. 

My versions:

- pdf-toolbox-document-0.0.6.0
- pdf-toolbox-content-0.0.4.0
- pdf-toolbox-core-0.0.4.0

**Comment by MaxDaten** [comment]
Hi, I investigated it a bit further. It seems to be an issue related to the OSX (10.11) Preview application. I trimmed the PDF with the viewer bei deleting other pages:

- Activating the thumbnail sidebar: `Alt+Cmd+2`
- Selecting pages and deleting with `backspace`

This pdf-toolkit seems to have some problems with pdfs manipulated this way. I tested [`pdf2txt`](https://github.com/euske/pdfminer/) with this problematic pdf and `pdf2txt` could parse the output correctly.

**Comment by Yuras** [comment]
Thank you for the report.

Does it work before deleting other pages? Can you share the original file too?

**Comment by MaxDaten** [comment]
I don't have any problems before deleting pages. I can't share the the original pdf in public, I will email it to your gmail address from your github profile page, if this is okay.

**Comment by Yuras** [comment]
OK, I'll investigate. Probably the original file is not necessary.

**Comment by Yuras** [comment]
OK, I found the reason. It is definitely a bug. Thank you again.

How urgent the issue is for you? If it is not, then I'll probably fix it during the weekend.

Technical details below, mostly for myself:

The `TT4` font has a unicode CMap with two `beginbfrange` sections:

```
50 beginbfrange
<21><21><0045>
<22><22><0073>
<23><23><0020>
<24><24><0069>
<25><25><0074>
<26><26><0064>
<27><27><0061>
<28><28><0065>
<29><29><0072>
...
<4e><4e><004e>
<4f><4f><004c>
<50><50><2013>
<51><51><0055>
<52><52><004b>
<53><53><003f>
endbfrange
5 beginbfrange
<54><54><0054>
<55><55><0041>
<56><56><0046>
<57><57><00df>
<58><58><0071>
endbfrange
```

Our CMap parser doesn't expect it, though the spec seems to allow that. As a result we are not able to decode glyphs from the second section.

**Comment by MaxDaten** [comment]
Hi and thanks for investigating this. It's not super urgent, so this weekend or one of the next is totally fine. I can work with the full document for a while.

**Comment by Yuras** [comment]
Fixed in stable branch via b07d0f25b0f980165a713f33dbd30c6fa921c712

I'll keep it open not to forget to apply it to master when/if I find time.

**Comment by Yuras** [comment]
Oh, and please check that it works for you. Then I'll upload it to hackage.

**Comment by MaxDaten** [comment]
Thanks for the fast response, I will test it tomorrow.

**Comment by MaxDaten** [comment]
Sorry for the delay. I tested it now with several stripped pdfs, and it works without problems. Thanks for the quick fix.

Another more general question about pdfs: I have another german pdf. When extracting the text via your `pdf-toolbox` or via `pdf2txt.py` some (not all) of the upper case letters are extracted as lower case. But in a pdf viewer the text is displayed correctly. I guess it is a problem with the pdf, any hints how to avoid or fix the pdf?

**Comment by Yuras** [comment]
Fixed in master via 87bb10a3e57b2c9cbedcf6aba2c328bbbbe9aae5

> Another more general question about pdfs

Sorry I missed the question initially.
It is not always possible to correctly extract text from PDF file -- often PDF file contains enough info to render glyphs, but not enough to determine what char they represent. It your case you can add (of fix if it presents, but is wrong) unicode cmap to the font, see `5.9.2 ToUnicode CMaps` section of the PDF spec.
`pdf-toolbox` has special module `Pdf.Toolbox.Core.Writer` for generating or editing PDF files, but it is low level.
